Change: make as many savegame chunks a CH_ARRAY as possible

TrueBrain opened the pull request:
## Motivation / Problem

Weird goal, I know right? But hear me out.

#9322 sets out to make to make our savegame self-descriptive (see motivation for that there). But in order to do that, we need to make all chunks a table.

The first step in making anything a table, is to first make it an array. And so we come to the motivation of this PR: to make that possible.

## Description

Most `CH_RIFF`s in fact are just an array of length 1. So it might seem silly at first, but that is exactly what this PR does: make those chunks into an array of length 1.

The exception was the `GLOG` chunk. That required some more love and attention.

After this PR, only the map-chunks are `CH_RIFF`. Those could be made into `CH_ARRAY`, but that is not really useful, as they cannot become self-describing for now anyway.

Lastly, the chunk table tells how chunks are stored on disk. Chunks that are no longer stored were still marked by their old type. This was often confusing. So I sneaked that change in here, to mark those as `CH_READONLY`, making it more clear that it is not by accident.

## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

TrueBrain committed:
Change: rework several CH_RIFF chunks to use CH_ARRAY instead

This adds two byte extra to those chunks, and might feel a bit
silly at first. But in later changes we will prefix CH_ARRAY with
a table header, and then this change shines.

Without this, we could still add headers to these chunks, but any
external reader wouldn't know if the CH_RIFF has them or not. This
way is much more practical, as they are now more like any other
chunk.

TrueBrain committed:
Change: rework GLOG chunk to be more like the rest

Basically it is very similar to Vehicles, where there first is
a type field, followed by data of that type. So this commit makes
it looks like how Vehicles solved that.
This removes a lot of custom "keeping track of length" stuff.

TrueBrain committed:
Codechange: mark chunks that are not stored as CH_READONLY

This makes it easier to spot chunks that have a save_proc that
is a nullptr, but also prevents confusion, where it looks like
the CH_ type of a chunk has influence on how it is being read.
It is not, it is only used for saving.

rubidium42 commented on src/saveload/animated_tile_sl.cpp:
```diff
@@ -18,13 +18,17 @@
 
 extern std::vector<TileIndex> _animated_tiles;
 
+static const SaveLoad _animated_tile_desc[] = {
```
In the commit message: practicle -> practical

rubidium42 commented on src/saveload/animated_tile_sl.cpp:
```diff
@@ -45,18 +49,25 @@ static void Load_ANIT()
-	_animated_tiles.resize(_animated_tiles.size() + count);
-	SlCopy(_animated_tiles.data(), count, SLE_UINT32);
+	if (IsSavegameVersionBefore(SLV_RIFF_TO_ARRAY)) {
+		uint count = (uint)SlGetFieldLength() / sizeof(_animated_tiles.front());
```
YIKYOMI, but count can be `size_t` as everything using it accepts `size_t`

rubidium42 commented on src/saveload/saveload.cpp:
```diff
@@ -1444,6 +1445,48 @@ static void SlDeque(void *deque, VarType conv)
+ * @param vector The std::vector being manipulated
+ * @param conv VarType type of variable that is used for calculating the size
+ */
+static void SlVector(void *vector, VarType conv)
```
I don't think it was meant that this got added in this commit.

TrueBrain replied:
It was, as this is the first use. I can split it off in a new commit that only adds it, but doesn't use it. Amount of code was low enough that I personally didn't care :)

TrueBrain committed:
Codechange: add the ability to save/load a std::vector

std::vector<bool> is not possible, as .. that is a nice special
case in C++.

This new type will be used in next commit.

rubidium42 commented on src/saveload/gamelog_sl.cpp:
```diff
@@ -15,147 +15,317 @@
+	void LoadCheck(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
 };
 
-static_assert(lengthof(_glog_desc) == GLCT_END);
```
Would it make sense to add something similar to `SlGamelogAction::description` but with a `+ 1`? Alternatively you could say that if someone adds a game log action but do not check it survives a save-load, then the game log action is pointless and should not be added.

TrueBrain replied:
Exactly the last part. The assert was there as the length of the list was important. That is no longer the case. And especially as you can now view a savegame, I think we have to assume the developer validates it :D
